ogzaf/demographics.py

```diff
@@ -366,44 +534,68 @@ def get_imm_resid(totpers, min_yr, max_yr):
             each period of life, length E+S
 
     """
-    pop_file = utils.read_file(
-        CUR_PATH, os.path.join("data", "demographic", "india_pop_data.csv")
+    pop_df = get_un_pop_data(start_year=2019, end_year=2021, download=False)
+    pop_2019 = (
+        pop_df["pop"][
+            (
+                (pop_df["sex_num"] == 3)
+                & (pop_df["year"] == 2019)
+                & (pop_df["age"] < 100)
+            )
+        ]
+        .to_numpy()
+        .flatten()
     )
-    pop_data = pd.read_csv(pop_file, encoding="utf-8")
-    pop_data_samp = pop_data[
-        (pop_data["Age"] >= min_yr - 1) & (pop_data["Age"] <= max_yr - 1)
-    ]
-    age_year_all = pop_data_samp["Age"] + 1
-    pop_2001, pop_2011 = (
-        np.array(pop_data_samp["2001"], dtype="f"),
-        np.array(pop_data_samp["2011"], dtype="f"),
-    )
-    pop_2001_EpS = pop_rebin(pop_2001, totpers)
-    pop_2011_EpS = pop_rebin(pop_2011, totpers)
-    # Create three years of estimated immigration rates for youngest age
-    # individuals
-    imm_mat = np.zeros((2, totpers))
-    fert_rates = get_fert(totpers, min_yr, max_yr, False)
-    mort_rates, infmort_rate = get_mort(totpers, min_yr, max_yr, False)
-    newbornvec = np.dot(fert_rates, pop_2001_EpS).T
-    # imm_mat[:, 0] = ((pop_2011_EpS[0] - (1 - infmort_rate) * newbornvec)
-    #                  / pop_2001_EpS[0])
-    imm_mat[:, 0] = 0
-    # Estimate immigration rates for all other-aged
-    # individuals
-    mort_rate10 = np.zeros_like(mort_rates[:-10])  # 10-year mort rate
-    for i in range(10):
-        mort_rate10 = mort_rates[i : -10 + i] + mort_rate10
-    mort_rate10[mort_rate10 > 1.0] = 1.0
-    imm_mat[:, 10:] = (
-        pop_2011_EpS[10:] - (1 - mort_rate10) * pop_2001_EpS[:-10]
-    ) / pop_2001_EpS[10:]
-    # Final estimated immigration rates are the averages over years
-    imm_rates = imm_mat.mean(axis=0)
-    neg_rates = imm_rates < 0
-    # For India, data were 10 years apart, so make annual rate
-    imm_rates = ((1 + np.absolute(imm_rates)) ** (1 / 10)) - 1
-    imm_rates[neg_rates] = -1 * imm_rates[neg_rates]
+    pop_2020 = (
+        pop_df["pop"][
+            (
+                (pop_df["sex_num"] == 3)
+                & (pop_df["year"] == 2020)
+                & (pop_df["age"] < 100)
+            )
+        ]
+        .to_numpy()
+        .flatten()
+    )
+    pop_2021 = (
+        pop_df["pop"][
+            (
+                (pop_df["sex_num"] == 3)
+                & (pop_df["year"] == 2021)
+                & (pop_df["age"] < 100)
+            )
+        ]
+        .to_numpy()
+        .flatten()
+    )
+    fert_rates = get_fert(totpers, min_yr, max_yr)
+    mort_rates, infmort_rate = get_mort(totpers, min_yr, max_yr)
```

Comment:
```suggestion
    fert_rates = get_fert(min_yr, max_yr)
    mort_rates, infmort_rate = get_mort(min_yr, max_yr)
```

Reply:
@SeaCelo. I need to work on this a little bit. We discussed hard coding the `totpers` object to be `100` and `min_yr=1` and `max_yr=100`. But I need to look back at all the tests and `calibrate.py` that use this to make sure this is OK. We might need it to be flexible.

Reply:
@SeaCelo. I removed the `min_yr` and `max_yr` arguments in `demographics.py` and `calibrate.py` as they are hard coded to `min_yr=1` and `max_yr=100` because the UN population data have annual ages. I left the `totpers` argument because we need to be able to transform the annual data to model periods that are bigger than one year (`totpers<100`).